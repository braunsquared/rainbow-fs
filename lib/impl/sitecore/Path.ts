import * as path from '../../util/path';
import { IPath } from '../../model/Sitecore';

export class Path implements IPath {
  readonly Name: string;
  readonly Folder: string;
  readonly Path: string;
  readonly Tokens: Array<string>;

  constructor(scPath: string) {
    const parsed = path.parse(scPath);
    this.Name = parsed.name;
    this.Folder = parsed.base;
    this.Path = parsed.path;
    this.Tokens = parsed.tokens;
  }

  rebase(oldRoot: string, newRoot: string): Path {
    // TODO: handle the fact that sitecore is not case sensitive
    if (!this.Path.startsWith(oldRoot)) {
      throw new Error('Unable to rebase when old root does not match');
    }

    return new Path(`${newRoot}/${this.Path.substring(oldRoot.length)}`);
  }

  relativePath(childPath: IPath): Path {
    if (!childPath.Path.startsWith(this.Path)) {
      throw new Error('Path is not a base of the child path');
    }

    return new Path(childPath.Path.substring(this.Path.length));
  }
}