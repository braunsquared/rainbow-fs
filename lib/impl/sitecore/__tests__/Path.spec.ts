import { Path } from '../Path';

describe('Path', () => {

  it('should parse a path into the appropriate segments', () => {
    const path = new Path('/sitecore/content/this/is/a/test');
    expect(path).toMatchSnapshot();
  });

  it('should throw an error when trying to rebase', () => {
    expect(() => {
      const path = new Path('/sitecore/content/this/is/a/test');
      path.rebase('/sitecore/content/not/a/base', '/site/content/new/base');
    }).toThrow();
  });

  it('should get a relative path', () => {
    const a = new Path('/sitecore/content/this');
    const b = new Path('/sitecore/content/this/is/a/test');
    expect(a.relativePath(b)).toMatchSnapshot();
  });

  it('should throw an error when trying to get a relative path', () => {
    expect(() => {
      const a = new Path('/sitecore/content/that');
      const b = new Path('/sitecore/content/this/is/a/test');
      a.relativePath(b);
    }).toThrow();
  });

});