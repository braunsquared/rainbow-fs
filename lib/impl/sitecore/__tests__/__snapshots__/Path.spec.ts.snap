// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path should get a relative path 1`] = `
Path {
  "Folder": "/is/a",
  "Name": "test",
  "Path": "/is/a/test",
  "Tokens": Array [
    "is",
    "a",
    "test",
  ],
}
`;

exports[`Path should parse a path into the appropriate segments 1`] = `
Path {
  "Folder": "/sitecore/content/this/is/a",
  "Name": "test",
  "Path": "/sitecore/content/this/is/a/test",
  "Tokens": Array [
    "sitecore",
    "content",
    "this",
    "is",
    "a",
    "test",
  ],
}
`;
